Source/WebKit/NetworkProcess/haiku/NetworkDataTaskHaiku.cpp

```diff
@@ -135,33 +166,168 @@ NetworkDataTask::State NetworkDataTaskHaiku::state() const
+
+        response.setHTTPStatusCode(statusCode);
+        response.setHTTPStatusText(httpResult->StatusText());
+
+        // Add remaining headers.
+        const BHttpHeaders& resultHeaders = httpResult->Headers();
+        for (int i = 0; i < resultHeaders.CountHeaders(); i++) {
+            BHttpHeader& headerPair = resultHeaders.HeaderAt(i);
+            response.setHTTPHeaderField(headerPair.Name(), headerPair.Value());
+        }
+
+        if (statusCode == 401) {
+        	
+        	//TODO
+        	
+            //AuthenticationNeeded((BHttpRequest*)m_request, response);
+            // AuthenticationNeeded may have aborted the request
+            // so we need to make sure we can continue.
+            
+            if (m_currentRequest.isNull())
+                return;
+        }
+    }
+
+    if (!m_client)
+        return;
+
+    if (m_redirected) {
+        m_redirectionTries--;
+
+        if (m_redirectionTries == 0) {
+            ResourceError error(url.host().utf8().data(), 400, url,
+                "Redirection limit reached");
+            
+            m_networkLoadMetrics.responseEnd = MonotonicTime::now() - m_startTime;
+            m_networkLoadMetrics.markComplete();
+            m_client->didCompleteWithError(error,m_networkLoadMetrics);
+            return;
+        }
+
+        // Notify the client that we are redirecting.
+        ResourceRequest request = m_currentRequest;
+        ResourceResponse responseCopy = response;
+        request.setURL(url);
+    	fprintf(stderr,"((((((( redirecting to %s)))))))\n",request.url().string().utf8().data());
+    	m_client->willPerformHTTPRedirection(WTFMove(responseCopy),WTFMove(request), [this](const ResourceRequest& newRequest){
+    		
+	    	if(newRequest.isNull() || m_state == State::Canceling)
+	    	return;
+	    	
+	    	m_startTime = MonotonicTime::now();//network metrics
+	    	
+	    	if( m_state != State::Suspended ){
+	    		m_state = State::Suspended;
+	    		resume();
+	    	}
+	    });
+    	
+    } else {
+        ResourceResponse responseCopy = response;
+        m_client->didReceiveResponse(WTFMove(responseCopy),[this](WebCore::PolicyAction policyAction){
+        	if(m_state == State::Canceling || m_state == State::Completed){
+        		return;
+        	}
+        	});
+    }
 }
 void NetworkDataTaskHaiku::DataReceived(BUrlRequest* caller, const char* data, off_t position,
         ssize_t size)
 {
-	if(size < 0 )
-	{
-		return;
-	}
-	m_client->didReceiveData(SharedBuffer::create(data,size));
+	if (m_currentRequest.isNull())
+        return;
+
+    if (!m_client)
+        return;
+
+    // don't emit the "Document has moved here" type of HTML
+    if (m_redirected)
+        return;
+
+    if (position != m_position)
+    {
+        debugger("bad redirect");
+        return;
+    }
+
+    if (size > 0) {
+        m_responseDataSent = true;
+        
+        
+        runOnMainThread([this,data = data, size = size]{
```

Comment:
You should review if these runOnMainThread are needed in the case of the network process. We can maybe avois a few jumps between threads and get better performance

Reply:
Hmm Curl people have done similar thing, anyways will try to look at it

Reply:
Yes, but if we use the BUrlProtocolAsynchronousListener, we are already in the main thread here.

Reply:
Ok then i think what should have been the main thread is not here a main thread ? things are in wrong place?

Reply:
If you use BUrlAsynchronousProtocolListener (I did not check if you do, but the WebKitLegacy code does), that already takes care of running the callbacks in the main thread. So, calling runOnMainThread from there is just the main thread asking itself to run things. There may be reasons to do this (it delays the execution to the next iteration of the message loop), but in this case I don't see why this would be needed, we could run the code directly.

If however you do not use the BUrlAsynchronousProtocolListener, but a standard BUrlProtocolListener, the callbacks will be called in the HTTP request thread. In that case, it is your responsibility to use runOnMainThread where needed to make sure WebKit internals code is not run in the HTTP request thread (since some parts of WebKit must run in one single thread, to greatly simplify thread locking problems).